### PR TITLE
backend/handlers: skip 0txs to mitigate address poisoning

### PR DESCRIPTION
**Pull request by Beerosagos**
Address Poisoning attack consists of an attacker creating empty (0 amount) txs to/from vanity addresses, hoping that the possible victim will copy/paste the wrong address from the transaction history when sending/receiving coins. This can happen in some ERC20 tokens, depending on the implementation of the `transfer` function inside the token contract (see
https://mirror.xyz/x-explore.eth/cL3d_CyNujXq8XY7ueP4omNXx_IY1EG5Dz0FD0vJ90M for details).

This hide 0 amount erc20 txs from the `transactions` endpoint, to mitigate the effects of this attack.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## Unreleased
 - Improve sats amount readability adding thousands separator
 - Update minimum supported buntu version to 20.04+
+- Hide 0 amount ERC20 txs to partially mitigate Address Poisoning Attack
```
Comment (benma) [inline CHANGELOG.md:6]: txs=>transactions. Imo no point in abbreviating where not necessary, as there are always people that don't know some acronyms or short names.

File: backend/coins/btc/handlers/handlers.go [hunk 2]
```diff
@@ -230,6 +231,10 @@ func (handlers *Handlers) getAccountTransactions(_ *http.Request) (interface{},
 	}
 	result.Transactions = []Transaction{}
 	for _, txInfo := range txs {
+		if txInfo.FeeIsDifferentUnit && big.NewInt(0).Cmp(txInfo.Amount.BigInt()) == 0 {
```
Comment (benma) [inline backend/coins/btc/handlers/handlers.go:234]: `FeeIsDifferentUnit` does not imply ERC-20 token necessarily. 

I think it's better to move this logic to `eth/account.go:Transactions()` (or to wherever the transactions are populated), where you can also directly check if it is an ERC-20 token. Generally/ideally, the HTTP handlers should not contain any logic but just parse input and assemble the JSON output.

**Review by benma — CHANGES_REQUESTED** [review]

File: backend/coins/btc/handlers/handlers.go (hunk 2 above)
Comment (Beerosagos) [reply backend/coins/btc/handlers/handlers.go:234]: As discussed, I didn't change this, to avoid modifying the persisted transactions, which could have consequences on e.g. balance computing. The code is only for frontend use anyway.

**Comment by Beerosagos** [comment]
@benma pushed an update. 
As agreed I didn't change the code in handlers: If we want to look for an alternative, think the easiest could be to modify `NewOrderedTransactions` (https://github.com/digitalbitbox/bitbox-wallet-app/blob/master/backend/accounts/transaction.go#L157), which is called by `Transactions()`, and making it returning a deep copy of the transaction list wihout 0 amount erc20 txs. lmk if you prefer this last approach.

File: backend/coins/btc/handlers/handlers.go (hunk 2 above)
Comment (benma) [reply backend/coins/btc/handlers/handlers.go:234]: The part about `FeeIsDifferentUnit does not imply ERC-20 token necessarily.` still applies though :) Best to make a new field in the tx-info struct `IsERC20` and set/use it accordingly

File: backend/coins/eth/account.go [hunk 3]
```diff
@@ -379,23 +379,44 @@ func (account *Account) update() error {
...
-		}
-		account.balance = coin.NewAmount(balance)
+		balance, err = account.coin.client.Balance(context.TODO(), account.address.Address)
+	}
+
+	if err != nil {
+		return errp.WithStack(err)
 	}
 
+	pendingAmount := pendingTxsAmount(outgoingTransactionsData, account.coin.erc20Token != nil)
```
Comment (benma) [inline backend/coins/eth/account.go:393]: seems like an unrelated commit slipped into this PR?

**Review by benma — CHANGES_REQUESTED** [review]

**Comment by Beerosagos** [comment]
@benma Added a new `IsErc20` field. The logic is pretty much the same as `FeeIsDifferentUnit`, I wonder if it would make more sense to just rename it, wdyt?

**Comment by benma** [comment]
> @benma Added a new `IsErc20` field. The logic is pretty much the same as `FeeIsDifferentUnit`, I wonder if it would make more sense to just rename it, wdyt?

No, because they are conceptually not equivalent. `FeeIsDifferentUnit` for example plays a role in the chart balance computation, and could be true for future tokens that are added which are not ERC-20.

File: backend/coins/btc/transactions/transactions.go [hunk 4]
```diff
@@ -492,6 +492,7 @@ func (transactions *Transactions) txInfo(
 		Size:             int64(txInfo.Tx.SerializeSize()),
 		Weight:           btcdBlockchain.GetTransactionWeight(btcutilTx),
 		CreatedTimestamp: txInfo.CreatedTimestamp,
+		IsErc20:          false,
```
Comment (benma) [inline backend/coins/btc/transactions/transactions.go:495]: can also leave it out like the other ETH fields - defaults to false anyway.

**Review by benma — APPROVED** [review]

**Comment by benma** [comment]
I merged this so it is part of the release candidate testing.

**Comment by Beerosagos** [comment]
> I merged this so it is part of the release candidate testing.

Cool, thanks!
